Haunted Blogの脆弱性の修正

jun-kondo opened the pull request.

jun-kondo committed:
ユーザーの入力文字列をエスケープする処理を追加

html_escapeメソッドを使用

jun-kondo committed:
SQLインジェクションの脆弱性を修正

クエリではプレースホルダを使用し入力値をそのまま受け取らないようにする

jun-kondo committed:
通常ユーザーがPremiumユーザーの機能を使える問題を修正

ユーザーの権限毎にストロングパラメーターを設定
Premiumユーザーのみrandom_eyecatchのパラメーターを受付可能に変更

jun-kondo committed:
他ユーザーのブログを変更・削除出来ないよう変更

編集、更新、削除は自分のブログのみ対象とする

jun-kondo committed:
他人の秘密のブログにアクセス出来てしまう問題を修正

ユーザー自身の全ブログ、全ての公開ブログのみアクセス出来るよう変更

jun-kondo committed:
Likeボタン押下時のJSON出力から個人情報(Email)を削除

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,7 +3,8 @@
   skip_before_action :authenticate_user!, only: %i[index show]
 
-  before_action :set_blog, only: %i[show edit update destroy]
+  before_action :set_blog, only: %i[show]
```
特定のアクションでしか実行しない処理で、かつそれが本質的な処理である場合は、フィルターで実行する必要はないと思います。フィルターを使うのは、アクション間で処理を共通化したい場合や、本質的でない処理をアクションから分離したい場合です。

jun-kondo replied:
フィルタの`set_blog`を削除しました。
d76a9753

処理を`show`アクションに移動しました
cfbfddf2

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -43,11 +44,23 @@ def destroy
 
   def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    if current_user.premium?
+      params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    else
+      params.require(:blog).permit(:title, :content, :secret)
+    end
```
`:random_eyecatch` 以外の部分が分岐間で重複しています。

jun-kondo replied:
`permit`に渡す属性値を配列で定義し、`premium?`が`true`だった場合`:random_eyecatch`を配列に追加するようにして重複を解消しました。

a90552cb

cafedomancer commented on app/helpers/blogs_helper.rb:
```diff
@@ -2,6 +2,6 @@
 module BlogsHelper
   def format_content(blog)
-    blog.content.gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
+    h(blog.content).gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
```
`rubocop:disalbe` が無くても RuboCop をパスするような実装を考えてください。

jun-kondo replied:
`simple_format`メソッドのhオプションを使用してエスケープと表示処理を実装するよう変更しました。

41c6bf93

cafedomancer commented on app/models/blog.rb:
```diff
@@ -7,10 +7,17 @@ class Blog < ApplicationRecord
   validates :title, :content, presence: true
 
-  scope :published, -> { where('secret = FALSE') }
+  scope :my_blogs, ->(user) { Blog.where(user_id: user.id) }
```
スコープ名には形容詞形を使ったほうがよいと思います。また、スコープ内で使用される `where` にクラスを指定してはいけません。スコープを連続して使用する場合に絞り込みが途中で解除されてしまいます。また、関連を指定するときは、ID ではなくオブジェクトそのものを指定するようにしましょう。

jun-kondo replied:
スコープ名をownedに変更しました。
また`id`ではなく`user`自体を指定するように変更しました。

c40eef1d

cafedomancer commented on app/models/blog.rb:
```diff
@@ -7,10 +7,17 @@ class Blog < ApplicationRecord
   scope :search, lambda { |term|
-    where("title LIKE '%#{term}%' OR content LIKE '%#{term}%'")
+    if term.present?
+      title_query, content_query = "%#{sanitize_sql_like(term)}%"
```
うーん、何かおかしいような...。

jun-kondo replied:
すみません。。。
同時代入の構文を間違えて書いていました。これだと2つ目の変数は常にnilになっていまいますね。。。

https://github.com/jun-kondo/haunted-blog/pull/1/commits/9691cc573a63df847d577e1dd0196105e1fcd7c6

cafedomancer commented on app/models/blog.rb:
```diff
@@ -7,10 +7,17 @@ class Blog < ApplicationRecord
+
+  scope :published, -> { where(secret: false) }
 
   scope :search, lambda { |term|
-    where("title LIKE '%#{term}%' OR content LIKE '%#{term}%'")
+    if term.present?
+      title_query, content_query = "%#{sanitize_sql_like(term)}%"
+      where('title LIKE ? OR content LIKE ?', title_query, content_query)
+    else
+      all
+    end
```
`term.present?` が `false` の場合に発行されるクエリが、変更前後で変わってしまっています。

jun-kondo replied:
`term`が`nil`のときは空文字を代入する処理を追加するようにしました。また`all`の実行部分を削除しました。

https://github.com/jun-kondo/haunted-blog/pull/1/commits/9691cc573a63df847d577e1dd0196105e1fcd7c6

cafedomancer commented on app/views/api/liking_users/_user.json.jbuilder:
```diff
@@ -1 +1 @@
-json.extract! user, :id, :email, :created_at, :updated_at, :nickname, :premium
+json.extract! user, :id, :created_at, :updated_at, :nickname, :premium
```
不要な情報だけを削除するのではなく、必要な情報のみを露出するほうが、考え方として安全だと思います。

jun-kondo replied:
`:id`と`:nickname`のみを露出するよう変更しました。

https://github.com/jun-kondo/haunted-blog/pull/1/commits/d3a3162f5c0d4bc209c37965061275b97013d7b4

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -43,11 +44,23 @@ def destroy
+
   def set_blog
-    @blog = Blog.find(params[:id])
+    @blog = if user_signed_in?
+              Blog.my_blogs(current_user).or(Blog.published).find(params[:id])
+            else
+              Blog.published.find(params[:id])
+            end
```
この条件分岐は無くせると思います。まずはノーヒントでチャレンジしてみてください。

jun-kondo replied:
分岐をなくして`Blog.where(user: current_user).or(Blog.where(secret: false).find(params[:id])`の結果を返すようにしました。
https://github.com/jun-kondo/haunted-blog/pull/1/commits/cfbfddf29fa9456d5cc559b4d57dae0b6a73502f

未ログインの場合、`current_user`が`nil`なので、クエリにはNULLが入ることになり`WHERE user_id IS NULL`ではなにも返さなくなります。そうすると、もう一つの条件の`WHERE secret" = false`の結果のみ返すので分岐省略できると思いましたが如何でしょうか？👀
```
SELECT "blogs".* FROM "blogs" WHERE ("blogs"."user_id" IS NULL OR "blogs"."secret" = ?) AND "blogs"."id" = ? LIMIT ?  [["secret", "[FILTERED]"], ["id", 2], ["LIMIT", 1]]
```

cafedomancer replied:
良さそうです！

jun-kondo committed:
simple_formatメソッドを使用

html_safeメソッドを使わずにHTMLタグを表示する

jun-kondo committed:
blog_paramsメソッドの修正

Premiumユーザーの場合、random_eyecatch属性をストロングパラメーターに追加して分岐時のコードの重複を解消する

jun-kondo committed:
scope :searchの修正

同時代入の構文が間違えていたので修正
termがnilのときは空文字を代入する処理を追加
sanitize_sql_likeの引数がnilだと例外が発生するから

jun-kondo committed:
scope :my_blogsをリネーム

:ownedにリネーム
idではなく関連モデルのuserを指定する

jun-kondo committed:
set_blogフィルタを削除

フィルタの対象がshowのみになったのでshowアクションに処理を移動した

jun-kondo committed:
showアクションでブログを取得する処理を追加

ユーザー自身の全ブログ、またはsecretステータスがfalseの全ブログからidと合致するものを検索する

jun-kondo committed:
JSON出力をidとnicknameのみに限定する

使用するの属性のみを露出するよう設定

jun-kondo commented:
修正、コメント返信致しましたー
再レビューよろしくお願い致します🙇‍♂️

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -43,11 +45,13 @@ def destroy
   def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    attributes = %i[title content secret]
+    attributes << [:random_eyecatch] if current_user.premium?
```
む、また何かおかしいような... (きちんとセルフレビューしましょう) 。

jun-kondo replied:
配列に配列を`push`して二重配列になっていますね。。。
失礼しました。シンボルを追加するように直しました🙇‍♂️

e2b09992

cafedomancer commented on app/models/blog.rb:
```diff
@@ -7,10 +7,15 @@ class Blog < ApplicationRecord
   validates :title, :content, presence: true
 
-  scope :published, -> { where('secret = FALSE') }
+  scope :owned, ->(user) { where(user:) }
```
`published` スコープは引数を取らないので published でよいですが、`owned` スコープは引数とをるのでスコープ名に何らかの前置詞を含めても良いような気がします。

jun-kondo replied:
`owned_by`にスコープの名称を変更しました。

58e84a2a

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -43,11 +45,13 @@ def destroy
+    @blog = current_user.blogs.find(params[:id])
   end
 
   def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    attributes = %i[title content secret]
+    attributes << [:random_eyecatch] if current_user.premium?
+    params.require(:blog).permit(attributes)
```
`permit` メソッドに渡している引数のデータ型が変更前後で変わってしまっています。

jun-kondo replied:
`permit(attributes)`を`permit(*attributes)`に変更して、配列を展開して渡すようにしました。

e2b09992

cafedomancer commented on app/models/blog.rb:
```diff
@@ -7,10 +7,15 @@ class Blog < ApplicationRecord
 
   scope :search, lambda { |term|
-    where("title LIKE '%#{term}%' OR content LIKE '%#{term}%'")
+    term = '' if term.nil? || term.strip.empty?
```
NilClass のメソッドを使えばもっとスッキリ書けるはずです。strip に関しては、脆弱性の修正のみを行うという観点に立つと不用意に元の挙動を変えてしまっていることになるので、今回は行うべきではないと思います。

jun-kondo replied:
> strip に関しては、脆弱性の修正のみを行うという観点に立つと不用意に元の挙動を変えてしまっていることになるので、今回は行うべきではないと思います。

`term`が`nil`になる状況を理解できておらず余計な処理を追加していました。`strip`の処理を削除しました。

> NilClass のメソッドを使えばもっとスッキリ書けるはずです。

`term.to_s`として、`nil`を`""`に変換するようにしました。

ba21c9cb

cafedomancer commented on app/models/blog.rb:
```diff
@@ -7,10 +7,15 @@ class Blog < ApplicationRecord
-    where("title LIKE '%#{term}%' OR content LIKE '%#{term}%'")
+    term = '' if term.nil? || term.strip.empty?
+    sanitized_term = sanitize_sql_like(term)
+    title_query = content_query = "%#{sanitized_term}%"
```
変数を分ける意味がないのでやめましょう。同じ変数名であれば同じ値を参照していることが分かりますが、違う変数名にすると違う値を参照しているかのような誤解が生じます。

jun-kondo replied:
承知しました。
クエリに代入する変数を1つに統一するように変更しました。

ba21c9cb

jun-kondo committed:
blog_paramsメソッドの修正

permitメソッドに渡す属性値が二重配列になっていたのをフラットな配列に変更
permitメソッドを呼び出す際、引数を可変長引数に記述を変更し配列を展開するように変更

jun-kondo committed:
scope :ownedを変更

スコープ名をowned_byに変更

jun-kondo committed:
scope :searchを変更

term == nilのときの対処: to_sメソッドでnilを空文字列に変換する
クエリに渡す変数を一つに統一: term_with_wildcardを使用する

jun-kondo commented:
修正、コメント返信致しましたー
再レビューよろしくお願い致します🙇‍♂️

cafedomancer commented on app/models/blog.rb:
```diff
@@ -7,10 +7,14 @@ class Blog < ApplicationRecord
-  scope :published, -> { where('secret = FALSE') }
+  scope :owned_by, ->(user) { where(user:) }
+
+  scope :published, -> { where(secret: false) }
```
コメントし忘れていましたが、このスコープの条件を Hash 形式に書き直しているのは良いですね！ 💡

jun-kondo replied:
> コメントし忘れていましたが、このスコープの条件を Hash 形式に書き直しているのは良いですね！ 💡

ありがとうございます😃

cafedomancer commented on app/models/blog.rb:
```diff
@@ -7,10 +7,14 @@ class Blog < ApplicationRecord
   scope :search, lambda { |term|
-    where("title LIKE '%#{term}%' OR content LIKE '%#{term}%'")
+    sanitized_term = sanitize_sql_like(term.to_s)
+    term_with_wildcard = "%#{sanitized_term}%"
```
この変数名だと sanitize されているという情報が失われてしまっていると思います。17 行目で `"%#{sanitized_term}%"` を 2 つ書くか、15 行目で `"%#{sanitized_term}%"` にする処理まで含めてやってしまうか、のどちらかにしたほうがよいと思います。

jun-kondo replied:
> 17 行目で "%#{sanitized_term}%" を 2 つ書くか、15 行目で "%#{sanitized_term}%" にする処理まで含めてやってしまうか、のどちらかにしたほうがよいと思います。

わかりました。論理的な連続性も重要なのですね。
こちら前者の内容で修正しました。

[9e70602](https://github.com/jun-kondo/haunted-blog/pull/1/commits/9e706024721a3a36dd29c5ec71d1e7fd9ee09926)

jun-kondo committed:
scope :searchを変更

term_with_wildcard変数を不使用しない記述に変更
sanitized_termをwhereメソッドの引数で使う

jun-kondo commented:
修正、コメント返信致しましたー
再レビューよろしくお願い致します🙇‍♂️
